### ISSUE TITLE
use of shared_ptr<const ProjData> etc still unsafe

### ISSUE THREAD

**Issue by KrisThielemans** [UCL/STIR#485]
We've tried to avoid giving access to internals of `ProjData` etc by only accepting `shared_ptr<const ProjData>` etc, see #470 #484  etc. However, this is *still unsafe* due to https://github.com/UCL/STIR/blob/b79bb323c1f1abab02f397f04ea8a13995b133c2/src/buildblock/ProjData.cxx#L400 and similar lines elsewhere.

Example
```c++
shared_ptr<ProjDataInfo>proj_data_info_sptr = initialise_somehow()
ProjDataInMemory p(proj_data_info_sptr, ....);
proj_data_info_sptr ->set_whatever();
```
Result: we've now modified a member in`p`, which is surprising to say the least, and will probably break use of `p` later on.

There isn't anything in `ProjData` that we can do against this, except for always cloning the `proj_data_info` object:
```c++
ProjData::ProjData(const shared_ptr<ExamInfo>& exam_info_sptr,
		   const shared_ptr<const ProjDataInfo>& proj_data_info_sptr)
  :ExamData(exam_info_sptr), proj_data_info_sptr(proj_data_info_sptr->create_shared_clone())
{}
```
Of course, this is the same for other classes, see e.g. 
https://github.com/UCL/STIR/blob/b79bb323c1f1abab02f397f04ea8a13995b133c2/src/include/stir/Sinogram.inl#L98-L105

Always cloning is however memory inefficient. It also means there's little or no point anymore in using `shared_ptr` at all.

Anyone any bright ideas? @rijobro?

**Comment by KrisThielemans** [comment]
Note that this will probably lead to quite surprising results in Python/MATLAB code, where a user could easily decide to re-use an existing object (which behind the scenes will be a shared_ptr).

**Comment by rijobro** [comment]
> Always cloning is however memory inefficient. It also means there's little or no point anymore in using shared_ptr at all.

I still think there's utility in having a shared pointer such that you can do `get_something() const` and not have to worry about the lifetime of the returned object.

> Result: we've now modified a member inp, which is surprising to say the least, and will probably break use of p later on.

However, this is very valid. Not sure what the right thing to do here is. Could we internally store a `unique_ptr`?

**Comment by KrisThielemans** [comment]
> I still think there's utility in having a shared pointer such that you can do get_something() const and not have to worry about the lifetime of the returned object.

agreed (although of course the user of that has to be aware that others can change the object pointed to)

> Could we internally store a unique_ptr?

a `unique_ptr` is unique. You cannot share it... I don't see the value in storing that. We could have the arguments as unique_ptr though. That way, you guarantee to the "constructor" that her own object will disappear. Therefore, she can no longer modify it! This seems to do the trick.

```c++
Sinogram(const Array<2,elemT>& p,  
          unique_ptr<ProjDataInfo >pdi_uptr,  ...) :
  proj_datA_info_sptr()pdi_uptr)
{...}
```
and
```c++
unique_ptr<ProjDataInfo>proj_data_info_uptr = initialise_somehow()l
ProjDataInMemory p(proj_data_info_uptr, ....);
proj_data_info_uptr ->set_whatever(); // should throw
```

What do you think?

Difficulties: 
- modifying all those classes again
- pre-C++11 compatibility (and potentially XCode nonsense). Our current work-around uses `std::auto_ptr`, which is broadly the same but not 100%.

**Comment by KrisThielemans** [comment]
I now realise that using `unique_ptr` probably resolves the problem but seems to get us into the problem of having to copy still quite a few times. e.g. `ProjData::get_viewgram()` needs to create a `Viewgram`, and will need to copy its `proj_data_info`. It seems the only safe option though.

Here's another example where things could explode currently
```c++
Viewgram viewgram = proj_data.viewgram(...);
// change num_tangential_poss somehow
viewgram.grow(...)
```
if `proj_data` and `viewgram` share a `proj_data_info`, it would break. We can prevent this by letting making our members `shared_ptr<const...>` such that the object cannot be modified (i.e. we'd have to create a clone and then modify). This is done in most/all of our recent PRs. However, we cannot give a guarantee to the user that this is what's going to happen. With `unique_ptr`, we would provide that guarantee I believe.

**Comment by rijobro** [comment]
And if the base level stored a `unique_ptr` and then passed `unique_ptr<const Obj>&` - that would be allowed wouldn't it? I'm guessing we wouldn't increase the use count, so could minimise the amount of data copied.

**Comment by KrisThielemans** [comment]
There seems no point in returning a `unique_ptr<const Obj>&`. Far better to just return a `const Obj&`.  It generally leads to more concise code anyway, and is what I've done in #487 (while still keeping `get_exam_info_sptr()` returning a `const` object for those functions who currently need a `shared_ptr<const Obj>` argument).

As you wrote above, the only reasons to return a pointer are to enable the caller to have an object that survives and/or can be shared.

Just to reiterate, the "`unique_ptr<Obj>` as argument" strategy avoids extra allocations in the case when `Obj` needs to be constructed first, as that object is now passed to the class that stores it (as in the example above). However, it still forces a clone when trying to re-use the result of a `get_obj_sptr` as argument to a constructor (or similar).

**Comment by KrisThielemans** [comment]
Potential problem: SWIG currently doesn't support `std::unique_ptr` https://github.com/swig/swig/issues/692 with a PR in progress https://github.com/swig/swig/pull/1722.
